### ISSUE TITLE
Don't fail the gradle task if the baseline file doesn't exist

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#3355]
Related with #3024 (this one is probably a prerequisite)

## Expected Behavior
If a baseline is configured in gradle but the file doesn't exist the task should not fail

## Current Behavior
It fails.

## Context
This is how we have it configured:
```kotlin
    @get:InputFile
    @get:Optional
    @get:PathSensitive(PathSensitivity.RELATIVE)
    val baseline: RegularFileProperty = project.objects.fileProperty()
```
and it seems that `@get:InputFile` forces the file to exist. And if it doesn't exist it fails. I tried to fix it but I know little about gradle.

I created the test already if someone want to use it: https://github.com/detekt/detekt/tree/baseline-doesnt-exist

**Comment by chao2zhang** [comment]
https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_input_output_validation Yes @InputFile requires the property to have a value and the file to exist.

Why would we want the task `detekt` to succeed? Do we expect users to run `detektBaseline` first?

**Comment by BraisGabin** [comment]
Each time that I create a new module I need to create the baselines files again (two because main and debug). A new module doesn't have debt so it should not need baseline.

And there's nearly no problem with miss configuration. If you miss configure your baseline you are going to find out it because of all the issues.

**Comment by chao2zhang** [comment]
We could use the following workaround:
```
    if (file("$projectDir/detekt/baseline.xml").exists()) {
        baseline = file("$projectDir/detekt/baseline.xml")
    }
```

**Comment by 3flex** [comment]
I like the way Android Lint handles baseline files - there's no task required to create baselines, instead it will create the baseline file if it's configured but the file doesn't exist yet, and reads it when running the lint task.

I've always thought this would be a good model for detekt to adopt. It would also help detekt-compiler-plugin remain "taskless", since there's no need for the `createBaseline` task.

**Comment by chao2zhang** [comment]
The concern is gradle cacheability, if the baseline file content gets changed (even we shouldn't, but sometimes it does), we should fail the up-to-date. Supporting this issue requires manually performing upToDate checks, which could be an overhead in the long run.

**Comment by schalkms** [comment]
I think the mentioned workaround from @chao2zhang is a good idea. Hence, I second that.

**Comment by cortinico** [comment]
> [docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_input_output_validation](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_input_output_validation) Yes @InputFile requires the property to have a value and the file to exist.

I believe our case is an exception. From the paragraph just below: https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_input_output_validation

```
You will occasionally want to disable some of this validation, specifically when an input file may validly not exist. That’s why Gradle provides the @Optional annotation: you use it to tell Gradle that a particular input is optional and therefore the build should not fail if the corresponding file or directory doesn’t exist.
```

Hence I belive we should be able to update the Gradle logic to follow the same as Android lint:
- No baseline specified => Do nothing with baselines
- Baseline specified AND file missing => Create the baseline (delegate to the `@TaskAction` of `detektBaseline`).
- Baseline specified AND file existing => Read and use the baseline.

**Comment by BraisGabin** [comment]
But we do have the `Optional` annotation already and the task fails before we can run our own code.

**Comment by cortinico** [comment]
> But we do have the `Optional` annotation already and the task fails before we can run our own code.

Seems like there is an open issue for exactly this on Gradle's end https://github.com/gradle/gradle/issues/2016

**Comment by BraisGabin** [comment]
Nice! There are some workaround there that I'll try. Thanks!!

**Comment by cortinico** [comment]
We should also consider changing `@get:InputFile` to `@get:Input` as we'll be using the Path rather than the file content for task avoidance.

**Comment by BraisGabin** [comment]
> We should also consider changing `@get:InputFile` to `@get:Input` as we'll be using the Path rather than the file content for task avoidance.

But, if the content of the baseline changes we want to rerun the task because the output of detekt can change, right? I think that we should keep using `@get:InputFile`

**Comment by cortinico** [comment]
> But, if the content of the baseline changes we want to rerun the task because the output of detekt can change, right?

Agree. Though we're forced to specify a valid file (we can't just create it if it's missing).

**Comment by 3flex** [comment]
I wonder if the baseline could be generated as a report and dropped in the reports directory? That would mean separate parameters are used for task input and output for the baseline file which makes it simple to solve the original issue.

If the file is generated by default the flow would be:
1. Run detekt
2. Copy baseline file from report output directory to a location of user's choice
3. Configure the input file location to be used for the check task(s)

There's nothing happening implicitly here which is nice in some respects but it would be awkward when the baseline changes often because of the manual copy/paste.

**Comment by cortinico** [comment]
> * Run detekt
> * Copy baseline file from report output directory to a location of user's choice
> * Configure the input file location to be used for the check task(s)

True, but this is as manual as:

- Run `detektBaseline`
- Configure the input file location to be used for the check task(s)
- Run `detekt`

That is our status quo so probably not really worth to pursue.

**Comment by 3flex** [comment]
> Configure the input file location to be used for the check task(s)

That depends on whether https://github.com/detekt/detekt/issues/1260 is addressed or not - if it is, this middle step could be removed (which is really another point in leaving the process as is, just thought I'd mention that other fairly stale issue).
